server/src-lib/Hasura/Eventing/ScheduledTrigger.hs

```diff
@@ -0,0 +1,392 @@
+  , seRetryConf     :: !(Param p RetryConfST)
+  }
+
+deriving instance Show (ScheduledEvent 'SE_PARTIAL)
+deriving instance Show (ScheduledEvent 'SE_FULL)
+
+type instance Param 'SE_PARTIAL a = ()
+type instance Param 'SE_FULL a = a
+
+-- empty splice to bring all the above definitions in scope
+$(pure [])
+
+instance ( J.ToJSON (Param p T.Text)
+         , J.ToJSON (Param p J.Value)
+         , J.ToJSON (Param p Int)
+         , J.ToJSON (Param p RetryConfST)
+         ) =>
+         J.ToJSON (ScheduledEvent p) where
+  toJSON = $(J.mkToJSON (J.aesonDrop 2 J.snakeCase) {J.omitNothingFields = True} ''ScheduledEvent)
+
+runScheduledEventsGenerator ::
+     L.Logger L.Hasura
+  -> Q.PGPool
+  -> IO SchemaCache
+  -> IO ()
+runScheduledEventsGenerator logger pgpool getSC = do
+  forever $ do
+    traceM "entering scheduled events generator"
+    sc <- getSC
+    let scheduledTriggers = Map.elems $ scScheduledTriggers sc
+    runExceptT
+      (Q.runTx
+         pgpool
+         (Q.ReadCommitted, Just Q.ReadWrite)
+         (insertScheduledEventsFor scheduledTriggers) ) >>= \case
+      Right _ -> pure ()
+      Left err ->
+        L.unLogger logger $ EventInternalErr $ err500 Unexpected (T.pack $ show err)
+    threadDelay oneHour
+
+insertScheduledEventsFor :: [ScheduledTriggerInfo] -> Q.TxE QErr ()
+insertScheduledEventsFor scheduledTriggers = do
+  currentTime <- liftIO getCurrentTime
+  let scheduledEvents = concatMap (generateScheduledEventsFrom currentTime) scheduledTriggers
+  case scheduledEvents of
+    []     -> pure ()
+    events -> do
+      let insertScheduledEventsSql = TB.run $ toSQL
+            SQLInsert
+              { siTable    = scheduledEventsTable
+              , siCols     = map (PGCol . T.pack) ["name", "scheduled_time"]
+              , siValues   = ValuesExp $ map (toTupleExp . toArr) events
+              , siConflict = Just $ DoNothing Nothing
+              , siRet      = Nothing
+              }
+      Q.unitQE defaultTxErrorHandler (Q.fromText insertScheduledEventsSql) () False
+  where
+    toArr (ScheduledEventSeed n t) = [(triggerNameToTxt n), (formatTime' t)]
+    toTupleExp = TupleExp . map SELit
+
+generateScheduledEventsFrom :: UTCTime -> ScheduledTriggerInfo-> [ScheduledEventSeed]
+generateScheduledEventsFrom time ScheduledTriggerInfo{..} =
+  let events =
+        case stiSchedule of
+          OneOff _ -> empty -- one-off scheduled events are generated during creation
+          Cron cron ->
+            generateSchedulesBetween
+              time
+              (addUTCTime nominalDay time)
+              cron
+   in map (ScheduledEventSeed stiName) events
+
+-- generates events (from, till] according to CronSchedule
+generateSchedulesBetween :: UTCTime -> UTCTime -> CronSchedule -> [UTCTime]
+generateSchedulesBetween from till cron = takeWhile ((>=) till) $ go from
+  where
+    go init =
+      case nextMatch cron init of
+        Nothing   -> []
+        Just next -> next : (go next)
```

Comment:
This could be written as `go = unfoldr (dup . nextMatch cron)`, where `dup` is from `Control.Arrow.Extended`.

Reply:
That line is quite something to understand :see_no_evil: . If you think it's standard, I can go ahead and make the change. Personally, I think it is quite complex and might become harder (for me) to maintain.

Reply:
I think `unfoldr` is fairly standard. The idea is that you provide two values: a “seed” of type `a`, and an “extrusion function” of type `a -> Maybe (b, a)`. The extrusion function is applied to the seed to produce two things: a value and a new seed. The value becomes the first element in the resulting list, and the new seed is fed back into the extrusion function to get obtain the next element. This process repeats until the extrusion function returns `Nothing`, at which points the list ends.

`unfoldr` is really the natural way to build a list from a function like `nextMatch`! This is because `unfoldr` captures the notion of “iterating” a function and collecting the results of each iteration into a list. It just so happens that, in this particular case, the seed value and the list elements are the same, so `dup` just says “use the same value for both.”

That said, I realized I wrote it slightly incorrectly: it should be `unfoldr (fmap dup . nextMatch cron)`, since you need `fmap` to go under the `Maybe`. I think that if you understand the individual pieces, it reads fairly nicely! But that’s admittedly subjective, so you can leave it as it is if you’d like.

Reply:
Yes, creating value and new seed is indeed what `nextMatch` seems to be doing.